### PR TITLE
Allow using CUDA-aware MPI

### PR DESCRIPTION
**Pull request by masterleinad**
@Rombur and me want to be able to use CUDA-awareness of the MPI implementation used if possible.
This is an early version where just the current functions in `Utilities::MPI` gain an additional `MemorySpace` template parameter. In case we have a CUDA-aware MPI implementation, we use basically the (normal) CPU implementation. Otherwise, we copy to the CPU, perform the operation there and copy back to the device.

Detecting CUDA-awareness is different for every implementation. Hence, we currently tend to let users set the `DEAL_II_WITH_CUDA_AWARE_MPI` flag themselves without any autodetection or check if this is safe to assume.

Are there any opinions on the current design before we port all the other relevant MPI functions to a version also working with CUDA data? The diff is best to be viewed without whitespace changes.

### CODE REVIEW

File: include/deal.II/base/mpi.h [hunk 1]
```diff
@@ -235,9 +237,12 @@ namespace Utilities
      * array is the sum over the i-th entries of the input arrays from each
      * processor.
      *
+     * This function can also be used for ArrayView objects initialized with
+     * data living in CUDA memory if MemorySpace::CUDA is chosen.
+     *
      * Input and output arrays may be the same.
      */
-    template <typename T>
+    template <typename T, typename MemorySpaceType = dealii::MemorySpace::Host>
```
Comment (bangerth) [inline include/deal.II/base/mpi.h]: I can't claim that I really understand what this is about, but how is this interpreted syntactically: that the object of type `T` is stored in device memory? That the function is called from the device? That the array view lives in the host memory space, but that the *pointer* that it stores points into device memory? (In which case the question is how one would access it.) Do you think you could clarify?

File: include/deal.II/base/mpi.h [hunk 2]
```diff
@@ -371,14 +379,35 @@ namespace Utilities
...
     void
     min(const ArrayView<const T> &values,
         const MPI_Comm &          mpi_communicator,
         const ArrayView<T> &      minima);
 
+    template <typename T, typename MemorySpaceType = dealii::MemorySpace::Host>
+    void
+    Isend(const ArrayView<const T> &values,
```
Comment (bangerth) [inline include/deal.II/base/mpi.h]: documentation missing; same for the next function

File: include/deal.II/base/mpi.h [hunk 3]
```diff
@@ -371,14 +379,35 @@ namespace Utilities
...
+          MPI_Datatype              datatype,
```
Comment (bangerth) [inline include/deal.II/base/mpi.h]: Can't you infer the data type from `T`? If that were possible, I think we ought to use this function everywhere because it would finally give us something more generic :-)  (But that would be for a separate patch.)

File: include/deal.II/base/mpi.h (hunk 2 above)
Comment (masterleinad) [reply include/deal.II/base/mpi.h]: Yes, that is all WIP.

File: include/deal.II/base/mpi.h (hunk 1 above)
Comment (masterleinad) [reply include/deal.II/base/mpi.h]: We assume that we have an `ArrayView` on the CPU that stores a pointer to data either on the CPU or on the GPU. The function is called from the CPU. In case the data is stored on the GPU, we can't access it directly.

File: include/deal.II/base/mpi.h (hunk 3 above)
Comment (masterleinad) [reply include/deal.II/base/mpi.h]: I guess template deduction already works with this version (in case we are using CPU memory) but I need to check.

**Comment by masterleinad** [comment]
Yes, I think storing information about the location of data pointed to in an `ArrayView` object makes sense. Currently, the class is agnostic to what it stores but accessing the data directly doesn't work if it doesn't live in CPU memory.

File: include/deal.II/base/mpi.h (hunk 3 above)
Comment (bangerth) [reply include/deal.II/base/mpi.h]: What I meant is that the `MPI_Datatype` seems redundant -- we already have a way to get that from the type `T`.

Comment (masterleinad) [reply include/deal.II/base/mpi.h]: Ah, sure. Let's keep that for a different PR.

**Comment by bangerth** [comment]
I guess the alternative might also be to invent a different class, say `CUDA::ArrayView` or something similar. Such a class might also be able to abstract the need for manual copying when necessary...

**Comment by Rombur** [comment]
cc: @dsambit @phanimotamarri

File: include/deal.II/base/mpi.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/base/mpi.h]: I think I mentioned this in another context already, but I think that's an abuse of the `ArrayView` class. If you can't see the data, then it's not a "view" -- it's really something else. I would prefer if there were a different class for that then, and in that case you also don't have to touch the existing MPI interfaces.

Comment (masterleinad) [reply include/deal.II/base/mpi.h]: At least for `LinearAlgebra::distributed::Vector` the challenge is to write code that both works for data living on the CPU and the GPU. Using `ArrayView` for both allows for code that doesn't need to be specialized too much. Of course, I could also just provide overloads for all the methods of the `Partitioner` class for raw pointers and use these. On the other hand, I try to avoid raw pointer interfaces as much as possible. If I try to represent an array having a pointer pointing to the first element and its length at hand, `ArrayView` seems to be a natural choice.

If were are storing the storage location inside the `VectorView` object, we don't need to touch existing MPI interfaces either (just the implementation). Having a different class name makes using the `MemorySpaceType` template parameter really ugly.

Would you like a different class that is essentially equal to `VectorView` but without an `operator[]` member function better?

For me, it is not too much a stretch to say that the data in `ArrayView` is accessible or not depending on the storage location of the data represented. Of course, I am biased here. What do others think?

Comment (Rombur) [reply include/deal.II/base/mpi.h]: ArrayView is `a class that represents a window of memory locations of type ElementType
and presents it as if it was an array that can be accessed via an operator[]` Which is exactly what we want. The only caveat is that the data lives on the device not on the host so you cannot access it from the host. That means that to use `operator[]` you need to be careful and store the address of the element in a pointer and then use this pointer in a kernel. So to access an element in the `ArrayView`, you need to do something like this:
```cuda
int *a = &my_array_view[2];
my_kernel<<<1,1>>>(a);
```
We could also just use a pair of a raw pointer and a size and use `using` to give it another name and hide the fact that it is a pair.

Comment (kronbichler) [reply include/deal.II/base/mpi.h]: I guess the question here depends on what an `ArrayView` is supposed to be in the context of the different memory spaces and whether, if I understand @bangerth correctly, it would be more appropriate to append the memory space to the `ArrayView` class rather than the functions in this context. Currently, `ArrayView` is only a wrapper around a pointer plus a size with some access checks, so the approach chosen by @masterleinad makes sense: The pointers do not carry any meaning about the context they are used in, it is the caller who needs to distinguish them.

I do like the fact that `ArrayView` is light-weight so I have a slight preference for what @masterleinad suggests. On the other hand, our concept is not really what e.g. `Kokkos::View` is, but I think that we would overengineer things if we tried to make `ArrayView` robust so that it allows the CPU to directly "look" into GPU memory (which means we copy from device to host in the background, which leads to either terribly slow or terribly complicated implementations).

**Comment by masterleinad** [comment]
The current state of this PR now passes the CUDA tests using CUDA-aware MPI for the `LinearAlgebra::distributed::Vector::update_ghost` methods.
We should be able to do the same for the `LinearAlgebra::distributed::Vector::compress` methods, but that might be done in a different PR.

What is left to do is to store the memory location inside the `ArrayView` object.

File: source/base/mpi.inst.in [hunk 4]
```diff
@@ -57,6 +57,30 @@ for (S : MPI_SCALARS)
...
+      MPI_Datatype,
+      int,
+      int,
+      MPI_Comm,
+      MPI_Request *);
+
+    template void Irecv<S, dealii::MemorySpace::CUDA>(
+      ArrayView<S>, MPI_Datatype, int, int, MPI_Comm, MPI_Request *);
+#endif
+
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply source/base/mpi.inst.in]: It's there. Just the formatting is different.

**Comment by masterleinad** [comment]
We just found out that we strictly must not use `DEAL_II_COMPILER_CUDA_AWARE` anywhere inside functions to decide about different behavior if compiled with `gcc` directly or with `nvcc`.
What we observe is that often times the `CPU` implementation is used in place of the `GPU` one.
We suspect that the symbol generated from the `nvcc` compile call is replaced by the `gcc` one instead of that both exist.
To avoid this problem, the easiest solution seems to be to templatize all these functions and only provide a specialization for the `nvcc` compile path. This way we are guaranteed to generate two symbols and should not run in troubles running the wrong code for CUDA files.

**Comment by masterleinad** [comment]
Trying to debug the non `MPI_WITH_CUDA_AWARE_MPI` case.

File: include/deal.II/base/partitioner.templates.h [hunk 5]
```diff
@@ -106,28 +104,50 @@ namespace Utilities
...
+                             chunk_size * sizeof(Number),
+                             cudaMemcpyDeviceToDevice);
+#    else
+                  static_assert(
+                    std::is_same<MemorySpaceType, MemorySpace::Host>::value,
+                    "");
+                  Assert(
+                    false,
+                    ExcMessage(
+                      "This function can only be called with MemeorySpace::CUDA "
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/partitioner.templates.h]: I realized that we really can't just hide everything in case we don't have `CUDA`-aware MPI. So that typo will vanish one way or the other. :wink:

**Comment by masterleinad** [comment]
This version should finally work both for `CUDA`-aware and non-`CUDA`-aware `MPI`.
I don't see a possibility to hide the different treatment of these two cases, since we need to keep the data to send and receive around. Of course, we could also just store the temporary host data needed for the non-`CUDA`-aware `MPI` case in the `Partitioner` class, but that would be against the current implementation that always assumes temporary storage to come from the outside.

This version only uses `CUDA`-aware-`MPI` for updating ghost values. The `compress` functions would be the next step.

**Comment by masterleinad** [comment]
Blocked by https://github.com/open-mpi/ompi/issues/6001

**Comment by masterleinad** [comment]
I implemented all the functionality I wanted apart from the `ArrayView` `MemorySpace` template parameter.

**Comment by masterleinad** [comment]
This is everything I wanted to do here. I will continue to extract some of the changes into separate PRs.

**Comment by dsambit** [comment]
Me and @phanimotamarri have started using the cuda_aware_mpi branch for gpu porting. So far we have been able to compile this branch with CUDA and CUDA_AWARE_MPI except the last few commits which now gives the following compilation error
```
/ccsopen/proj/gen109/GPAWU/software/fordftfe/dealii/dealiiCUDA/include/deal.II/lac/cuda_kernels.h(131): error: no operator ">" matches these operands
            operand types are: const std::complex<double> > const std::complex<double>
          detected during:
            instantiation of "Number dealii::LinearAlgebra::CUDAWrappers::kernel::Binop_Min<Number>::operation(Number, Number) [with Number=std::complex<double>]" 
/ccsopen/proj/gen109/GPAWU/software/fordftfe/dealii/dealiiCUDA/include/deal.II/lac/cuda_kernels.templates.h(57): here
            instantiation of "void dealii::LinearAlgebra::CUDAWrappers::kernel::vector_bin_op<Number,Binop>(Number *, const Number *, dealii::LinearAlgebra::CUDAWrappers::kernel::size_type) [with Number=std::complex<double>, Binop=dealii::LinearAlgebra::CUDAWrappers::kernel::Binop_Min]"
```
We have found a workaround by using -DDEAL_II_WITH_COMPLEX_VALUES = OFF.

**Comment by dsambit** [comment]
@masterleinad Thank you fixing the compilation issue in complex case.
We are now facing a performance issue related to the GPU ported update_ghost_values and compress.
The FEM mesh we are testing has 2,588,505 dofs distributed across 6 GPUs and we use 6 MPI tasks.
**Our algorithm essentially calls update_ghost_values and compress inside a loop of size 200**.
What we observed is that using 1 MPI task we took 6 sec and using 6 MPI tasks we took around 40 sec. Next we profiled the run using nvprof and identified the following hotspots and the plausible reasons for them:  

**(compress related kernel calls and API calls and associated costs)**
700,000 kernel calls to dealii::LinearAlgebra::CUDAWrappers::kernel::vector_bin_op which took 2.3 sec.
700,000 API calls to cudaLaunchKernel took 10 sec. (**kernel launch overhead**)

Plausible reason for so many kernel calls in case of compress:
for loop over import_ranges
```
              for (const auto &import_range : import_indices_data)
                {
                  const auto chunk_size =
                    import_range.second - import_range.first;
                  const int n_blocks =
                    1 + (chunk_size - 1) / (::dealii::CUDAWrappers::chunk_size *
                                            ::dealii::CUDAWrappers::block_size);
                  dealii::LinearAlgebra::CUDAWrappers::kernel::vector_bin_op<
                    Number,
                    dealii::LinearAlgebra::CUDAWrappers::kernel::Binop_Addition>
                    <<<n_blocks, dealii::CUDAWrappers::block_size>>>(
                      locally_owned_array.data() + import_range.first,
                      read_position,
                      chunk_size);
                  read_position += chunk_size;
                }
```

**(update_ghost_values related kernel calls and API calls and associated costs)**
700,000 kernel calls to [CUDA memcpy DtoD] which took 1 sec.
700,000  API calls to cudaMemcpy which took 22 sec. (**CUDA memcpy launch overhead**)

Plausible reason for so many CUDA memcpy calls in case of update_ghost_values:
for loop over import_targets
```
      for (unsigned int i = 0; i < n_import_targets; i++)
        {
          // copy the data to be sent to the import_data field
          std::vector<std::pair<unsigned int, unsigned int>>::const_iterator
            my_imports = import_indices_data.begin() +
                         import_indices_chunks_by_rank_data[i],
            end_my_imports = import_indices_data.begin() +
                             import_indices_chunks_by_rank_data[i + 1];

          for (; my_imports != end_my_imports; ++my_imports)
            {
              const unsigned int chunk_size =
                my_imports->second - my_imports->first;
#    if defined(DEAL_II_COMPILER_CUDA_AWARE) && \
      defined(DEAL_II_WITH_CUDA_AWARE_MPI)
              if (std::is_same<MemorySpaceType, MemorySpace::CUDA>::value)
                {
                  const cudaError_t cuda_error_code =
                    cudaMemcpy(temp_array_ptr + index,
                               locally_owned_array.data() + my_imports->first,
                               chunk_size * sizeof(Number),
                               cudaMemcpyDeviceToDevice);
                  AssertCuda(cuda_error_code);
                }
              else
#    endif
                {
                  std::memcpy(temp_array_ptr + index,
                              locally_owned_array.data() + my_imports->first,
                              chunk_size * sizeof(Number));
                }
              index += chunk_size;
            }
          AssertDimension(index, import_targets_data[i].second);

          // start the send operations
          const int ierr =
            MPI_Isend(temp_array_ptr,
                      import_targets_data[i].second * sizeof(Number),
                      MPI_BYTE,
                      import_targets_data[i].first,
                      my_pid + communication_channel,
                      communicator,
                      &requests[n_ghost_targets + i]);
          AssertThrowMPI(ierr);
          temp_array_ptr += import_targets_data[i].second;
        }
```
**Some thoughts on a possible optimization strategy**
I am wondering if there is a way to reduce the number of kernel calls, for example in case of compress by using a binning approach, which collects import_ranges which do not intersect with each other into a single bin and then call one single kernel call to dealii::LinearAlgebra::CUDAWrappers::kernel::Binop_Addition for that bin. In that way, the number of kernel calls will be equal to the number of bins.

**Comment by masterleinad** [comment]
Thanks for profiling! Can you provide your test case?

**Comment by dsambit** [comment]
Sure, I will create the test case right away.

**Comment by dsambit** [comment]
I have attached the test case which I ran on 6 gpus on a single node with 6 MPI tasks.
Following is my output. GPU call is 420 times slower than the CPU call.
I have also put the nvprof output in the attached folder.
```
Number of Devices 6
Device Id currently used is 4 for taskId: 4
Device Id currently used is 3 for taskId: 3
Device Id currently used is 5 for taskId: 5
Device Id currently used is 1 for taskId: 1
Device Id currently used is 2 for taskId: 2
Device Id currently used is 0 for taskId: 0
number of elements: 32768
number of degrees of freedom: 8586756
Time on CPU for 200 update_ghost_values and compress calls: 0.0506373 sec
Time on GPU for 200 update_ghost_values and compress calls: 21.276 sec
```

[dealiiCUDAMPIProfiling.zip](https://github.com/dealii/dealii/files/2546362/dealiiCUDAMPIProfiling.zip)

**Comment by kronbichler** [comment]
The part of the code that really hurts is an optimization for the CPU: In order to avoid storing a separate index for every item we need to exchange, ranges of indices are put together. Of course this cannot work on the GPU and we should simply list all ghost indices in an array. I think `Partitioner` could hold both a compressed array as we have today (for consumption in a CPU) as well as a plain array of all indices which would allow us to collect data as
```
  tmp_array[i] = locally_owned_array[import_indices_plain[i]];
```
Such a loop can then be fed to CUDA and should be reasonably efficient.

**Comment by dsambit** [comment]
>  `tmp_array[i] = locally_owned_array[import_indices_plain[i]];`
> Such a loop can then be fed to CUDA and should be reasonably efficient.

I think this approach should work for update_ghost_values on GPU, but for compress
this can lead to a race condition if different threads are doing operations on the same location of the array.

**Comment by kronbichler** [comment]
> for compress this can lead to a race condition 

This is correct, but we could try to use `LinearAlgebra::CUDAWrappers::atomicAdd_wrapper` on all elements. As far as I remember, this should be reasonably well-performant. Alternatively, we could use your idea of bins as well - I'm not sure which one would be better, and did not think about that way of structuring the loop which avoids the conflicts (but it could be done).

**Comment by Rombur** [comment]
> This is correct, but we could try to use LinearAlgebra::CUDAWrappers::atomicAdd_wrapper on all elements. As far as I remember, this should be reasonably well-performant.

Yes, newer GPUs are very good at doing atomics. I think this will be simpler to implement so that's what I would try first.

**Comment by dsambit** [comment]
@masterleinad I am definitely interested to contribute here, once I resolve some current issues we have with regards to gpu porting of our DFT-FE code.

**Comment by Rombur** [comment]
@dsambit I have modified `update_ghost_values` and it does go faster on your benchmark but not by much. The reason is that on my machine the very first call to `MPI_Isend` is `10^5` slower than the following ones! This means that it totally dominates everything thing else (the first `MPI_Isend` is several times slower than the following 200 `update_ghost_values`...). Do you observe something similar on your machine? If I discard the first `MPI_Isend` the CPU vs GPU comparison is not as bad. The new implementation does help a lot by decreasing the number of kernel launch but it is still slower than on the CPU implementation.

**Comment by dsambit** [comment]
@Rombur Regarding the testing of the above update_ghost_values optimization, unfortunately cuda aware MPI currently has a problem on our machine. I will do the tests as soon that issue is fixed.

**Comment by masterleinad** [comment]
I guess this issue is not directly related to this PR. In particular, I assume that you should see this problem with every test `cuda` test in the testsuite on the `master` branch. Can you check this via
 ```
make setup_tests_cuda && ctest -V -R "cuda"
```
?

**Comment by dsambit** [comment]
@masterleinad your hunch is correct. The same issue is also present in the master branch- I checked by compiling the minimal example.

> Can you check this via
> make setup_tests_cuda && ctest -V -R "cuda"
Still trying to set this up as the machine I have access to has GPUs only on compute nodes.

**Comment by masterleinad** [comment]
Then, it would be best to open a separate issue for the problem.

**Comment by dsambit** [comment]
I did some performance tests on the optimized update_ghost_values using multiple P100 GPUs. I am using a slightly modified test (see [cudaMPIProfileNew.zip](https://github.com/dealii/dealii/files/2726269/cudaMPIProfileNew.zip)). Here instead of a single component parallel vector used in the earlier test, I am using a multi-component vectors with the number of components specified by a block size. Still the vector is stored as a single parallel vector- the components for each degree of freedom are stored contiguously in a flattened format allowing for the data for all components to be communicated simultaneously. This test replicates the use of udpate_ghost_values in our application. Below are the performance results (I do not time the first call to update_ghost_values as suggested by @Rombur )

```
For block size=400 (parallel vector with 400 components) on
 16 P100 GPUs using 16 tasks (4 tasks per node)
number of elements: 4096
number of degrees of freedom: 1098500
Time on CPU for 200 update_ghost_values calls: 2.29 sec
Time on GPU for 200 update_ghost_values calls: 3.28 sec
```
This is clearly satisfactory for our needs.

However, for block size=1 (parallel vector with single component), update_ghost_values on GPU is 20 times slower than the CPU counterpart, which is still a big improvement over being 400 times slower earlier. I think the reason for this that between block size=400 and block size=1, block size=400 takes advantage of coalesced memory access (all 400 components are contiguous for a given dof) in the gather kernel.

**Comment by masterleinad** [comment]
Let's move on with this PR then. We can always improve performance later. I will squash my commits.
